PR: https://github.com/odoo/odoo/pull/263225

From: 064aad50686600f236044964d30224f7bb4b9b48
From: FedericoBraidi
Date: 2026-05-28 09:25:53

Breaking data model changes scores: del:13 + add:0, change matches:
-    period_start = fields.Date('Period Start', compute='_compute_period_dates', store=True, readonly=False)
-    period_stop = fields.Date('Period Stop', compute='_compute_period_dates', store=True, readonly=False)
-    export_file = fields.Binary('Export File', readonly=True)
-    export_filename = fields.Char('Export Filename', readonly=True)
-    company_id = fields.Many2one(
-    eligible_employee_line_ids = fields.One2many(
-    eligible_employee_count = fields.Integer(
-    creation_date = fields.Date(
-    create_uid = fields.Many2one('res.users', index=True)
-    export_id = fields.Many2one('hr.export.mixin', required=True, index=True, ondelete='cascade')
-    company_id = fields.Many2one(related='export_id.company_id', store=True, readonly=True)
-    employee_id = fields.Many2one('hr.employee', required=True, ondelete='cascade', check_company=True)

Total Changes: 107

[IMP] hr_work_entry: introduce unified work entry export

We want to unify the way work entries are exported from Odoo. Currently,
we have different modules for each of the most important 3rd party
applications that accept work entry reports but here we want to drop
those and only create one work entry export format for Odoo.

The exports will be accessible from the employee views via the cog menu
or the action menu. In hr_presence we unintelligently override the
whole cog menu/action menu of the views to add the presence-related
buttons. This is a problem because depending on the installation order
of the modules we might lose the content of one or the other. To avoid
this, here we instead create a common override in hr and then add to it
in both hr_work_entry and hr_presence.

closes odoo/odoo#263225

Task: 6196138
Related: odoo/enterprise#116712
Related: odoo/upgrade#10175
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/__init__.py
+++ b/addons/hr/models/__init__.py
@@ -28,4 +28,3 @@ from . import resource
 from . import resource_calendar
 from . import resource_calendar_leaves
 from . import ir_ui_menu
-from . import hr_export_mixin

--- a/addons/hr/models/hr_export_mixin.py
+++ b/None
@@ -1,106 +0,0 @@
-from dateutil.relativedelta import relativedelta
-
-from odoo import api, fields, models
-from odoo.exceptions import UserError
-from odoo.fields import Domain
-from odoo.tools import BinaryBytes
-
-
-class HrExportMixin(models.AbstractModel):
-    _name = 'hr.export.mixin'
-    _description = 'HR Export Mixin'
-
-    @api.model
-    def _country_restriction(self):
-        return False
-
-    @api.model
-    def default_get(self, fields):
-        country_restriction = self._country_restriction()
-        if country_restriction and country_restriction not in self.env.companies.mapped('country_id.code'):
-            raise UserError(self.env._(
-                'You must be logged in a %(country_code)s company to use this feature',
-                country_code=country_restriction
-            ))
-        return super().default_get(fields)
-
-    def _get_company_domain(self):
-        domain = Domain('id', 'in', self.env.companies.ids)
-        if restriction := self._country_restriction():
-            domain &= Domain('partner_id.country_id.code', '=', restriction)
-        return domain
-
-    period_start = fields.Date('Period Start', compute='_compute_period_dates', store=True, readonly=False)
-    period_stop = fields.Date('Period Stop', compute='_compute_period_dates', store=True, readonly=False)
-    export_file = fields.Binary('Export File', readonly=True)
-    export_filename = fields.Char('Export Filename', readonly=True)
-    company_id = fields.Many2one(
-        'res.company', domain=lambda self: self._get_company_domain(),
-        default=lambda self: self.env.company, required=True)
-    eligible_employee_line_ids = fields.One2many(
-        'hr.export.employee.mixin', 'export_id',
-        string='Eligible Employees')
-    eligible_employee_count = fields.Integer(
-        '#Employees', compute='_compute_eligible_employee_count')
-    creation_date = fields.Date(
-        string="Creation Date",
-        compute="_compute_creation_date",
-        store=False,
-    )
-    create_uid = fields.Many2one('res.users', index=True)
-
-    @api.depends('eligible_employee_line_ids')
-    def _compute_eligible_employee_count(self):
-        for export in self:
-            export.eligible_employee_count = len(export.eligible_employee_line_ids)
-
-    def _compute_period_dates(self):
-        for export in self:
-            export.period_start = fields.Date.today().replace(day=1)
-            export.period_stop = export.period_start + relativedelta(months=1, days=-1)
-
-    def _compute_creation_date(self):
-        for rec in self:
-            rec.creation_date = rec.create_date.date() if rec.create_date else False
-
-    def action_export_file(self):
-        self.ensure_one()
-        self._check_before_export()
-        file_content = self._generate_export_file()
-        filename = self._generate_export_filename()
-        self.write({
-            "export_file": BinaryBytes(file_content),
-            "export_filename": filename,
-        })
-
-    def _check_before_export(self):
-        return True
-
-    def _generate_export_file(self) -> bytes:
-        raise NotImplementedError()
-
-    def _generate_export_filename(self):
-        raise NotImplementedError()
-
-    def action_open_employees(self):
-        self.ensure_one()
-        return {
-            'name': self.env._("Eligible Employees"),
-            'type': "ir.actions.act_window",
-            'res_model': self.eligible_employee_line_ids._name,
-            'view_mode': "list",
-            'domain': [('id', 'in', self.eligible_employee_line_ids.ids)],
-            'context': dict(self.env.context, default_export_id=self.id),
-        }
-
-    def action_populate(self):
-        raise NotImplementedError()
-
-
-class HrExportEmployeeMixin(models.AbstractModel):
-    _name = 'hr.export.employee.mixin'
-    _description = 'HR Export Employee'
-
-    export_id = fields.Many2one('hr.export.mixin', required=True, index=True, ondelete='cascade')
-    company_id = fields.Many2one(related='export_id.company_id', store=True, readonly=True)
-    employee_id = fields.Many2one('hr.employee', required=True, ondelete='cascade', check_company=True)
